tools/test/h5gentest.c

```diff
@@ -0,0 +1,607 @@
+    gent_vlenstr_array();
+    gent_char();
+    gent_attr_all();
+    gent_compound_complex();
+    gent_compound_complex2();
+    gent_named_dtype_attr();
+    gent_null_space();
+    gent_zero_dim_size();
+
+    gent_filters();
+    gent_fvalues();
+    gent_fcontents();
+    gent_string();
+    gent_aindices();
+    gent_longlinks();
+    gent_ldouble();
+    gent_ldouble_scalar();
+    gent_binary();
+    gent_bigdims();
+    gent_hyperslab();
+    gent_group_creation_order();
+    gent_attr_creation_order();
+    gent_fpformat();
+    gent_extlinks();
+    gent_fs_strategy_threshold();
+    gent_packedbits();
+    gent_dataset_idx();
+    gent_attr_intsize();
+    gent_charsets();
+    gent_compound_intsizes();
+    gent_compound_attr_intsizes();
+
+    gent_nested_compound_dt();
+    gent_intscalars();
+    gent_attr_intscalars();
+    gent_string_scalars();
+    gent_compound_int_array();
+    gent_compound_ints();
+    gent_intattrscalars();
+    gent_intsattrs();
+    gent_floatsattrs();
+    gent_bitnopaquefields();
+    gent_nodata();
+
+    gent_intsfourdims();
+    gent_null_space_group();
+
+    gent_udfilter(H5DUMP_UDFILTER_FILE);
+
+    gent_err_attr_dspace();
+
+    /* Generate the files for testing Onion VFD */
+    gent_onion_1d_dset();
+    gent_onion_create_delete_objects();
+    gent_onion_dset_extension();
+
+#ifdef H5_HAVE__FLOAT16
+    gent_float16();
+    gent_float16_be();
+#endif
+
+#ifdef H5_HAVE_COMPLEX_NUMBERS
+    gent_complex();
+    gent_complex_be();
+#endif
+}
+
+static void
+gen_h5fc_files(void)
+{
+    unsigned i, new_format;
+
+    /* Generate a non-latest-format file with v3 superblock */
+    gen_non(NON_V3_FILE);
+
+    /* Generate a new format file with a no-filter-edge-chunk dataset */
+    gen_edge(EDGE_V3_FILE);
+
+    /* Generate a new format file with 'K' value of 1 in H5Pset_istore_k() */
+    gen_err_level(ERR_LEVEL_FILE);
+
+    /* Generate old/new format file with/without messages in the superblock extension */
+    for (new_format = false; new_format <= true; new_format++) {
+        for (i = 0; i < 8; i++) {
+            char filename[50];
+
+            memset(filename, 0, sizeof(filename));
+            if (!new_format)
+                strcat(filename, "old_");
+            strcat(filename, H5FC_FILENAME[i]);
+
+            gen_ext(filename, new_format, i);
+        } /* end for */
+    }     /* end for */
+}
+
+static void
+gen_h5jam_files(void)
+{
+    if (create_textfile(UBTXT2, 10) < 0)
+        goto error;
+    if (create_textfile(UBTXT3, 511) < 0)
+        goto error;
+    if (create_textfile(UBTXT4, 512) < 0)
+        goto error;
+    if (create_textfile(UBTXT5, 513) < 0)
+        goto error;
+
+    if (gent_ub(H5JAM_FILE7, 0, 0) < 0)
+        goto error;
+    if (gent_ub(H5JAM_FILE8, 512, PATTERN_LEN) < 0)
+        goto error;
+    if (gent_ub(H5JAM_FILE9, 1024, 513) < 0)
+        goto error;
+
+    return;
+error:
+    fprintf(stderr, "h5jam test generator FAILED\n");
```

Comment:
May want to make the main generator functions for each tool return some sort of value so that can at least be captured when failures happen.